File: android/src/main/kotlin/com/syncrhrohealth/bandpass_ble_connection/core/handler/DeviceHandler.kt
package com.syncrhrohealth.bandpass_ble_connection.core.handler

import android.bluetooth.BluetoothDevice
import android.content.Context
import android.util.Log
import com.syncrhrohealth.bandpass_ble_connection.ble.BleConnectionCallBack
import com.syncrhrohealth.bandpass_ble_connection.ble.BlePeripheral
import com.syncrhrohealth.bandpass_ble_connection.core.parser.DataParser
import com.syncrhrohealth.bandpass_ble_connection.util.ThreadPoolCreater

class DeviceHandler(
    context: Context,
    device: BluetoothDevice,
    val callBack: DeviceHandlerCallBack
) :
    BleConnectionCallBack {

    private val peripheral: BlePeripheral = BlePeripheral(context, device, this)

    private var parser: DataParser = DataParser(this) {
        ReceiveDataHandler.handle(it, this)
    }

    val executor = ThreadPoolCreater.create()

    init {
        executor.submit(parser)
    }

    override fun onConnected(connection: BluetoothDevice) {
        callBack.onConnected(this)
    }

    override fun onDisconnected(connection: BluetoothDevice) {
        callBack.onDisConnected(this)
    }

    override fun onDataReceived(connection: BluetoothDevice, byteArray: ByteArray) {
        parser.push(byteArray)
    }

    override fun onBatteryLevelReceived(connection: BluetoothDevice, byteArray: ByteArray) {
        // Handle battery level data if needed
        Log.e(TAG, "onBatteryLevelReceived: array: $byteArray")
        peripheral.batteryLevel = byteArray[0].toInt()
        Log.e(TAG, "Battery Level: ${peripheral.batteryLevel}")
        callBack.onBatteryLevelRsp(
            peripheral.batteryLevel,
            this
        )
    }

    fun requestToConnect(autoConnect: Boolean = false) {
        peripheral.requestToConnect(autoConnect)
    }

    fun reconnect() {
        peripheral.reconnect()
    }

    fun requestToDisconnect() {
        peripheral.requestToDisconnect()
    }

    fun isEqual(address: String): Boolean {
        return peripheral.getAddress() == address
    }

    fun getDevice(): BlePeripheral {
        return peripheral
    }

    fun write(data: ByteArray) {
        peripheral.write(data)
    }
}

File: ios/Classes/Core/Handler/DeviceHandler.swift
//
//  DeviceHandler.swift
//  Runner
//
//  Created by MAC on 11/3/25.
//

import Foundation
import CoreBluetooth

class DeviceHandler: DataParserCallback, BleConnectionToDeviceCallback {
  
    private var peripheral: BlePeripheral?
    private var callbackToCentral: DeviceHandlerToCentralCallback?
    private var parser: DataParser?
    
    init(peripheral: CBPeripheral, callback: DeviceHandlerToCentralCallback, central: BleCentralManager) {
        self.peripheral = BlePeripheral(peripheral: peripheral, callback: self, central: central)
        self.callbackToCentral = callback
        
        parser = DataParser(callback: self)
        parser?.start()
    }
    
    func getCallBackToCentral() -> DeviceHandlerToCentralCallback? {
        return callbackToCentral
    }
    
    func connect() {
        peripheral?.connect()
        
    }
    
    func getConnection() -> BleConnection? {
        peripheral?.getConnection()
    }
    
    func changeDisconnectStatus() {
        peripheral?.changeDisconnectStatus()
    }
    
    func requestToDisconnect() {
        peripheral?.requestToDisconnect()
        parser?.stop()
    }
    
    func write(data: Data) {
        peripheral?.write(data: data)
    }
    
    func getDevice() -> BlePeripheral? {
         return peripheral
    }
    
    func getCBPeripheral() -> CBPeripheral? {
        return peripheral?.getDevice()
    }
    
    //************************** DATA PARSER CALLBACK *************************************************
    
    func onPacketRecieved(packet: Packet) {
        ReceiveDataHandler.handle(packet: packet, handler: self)
    }
    
    //************************** BLE CONNECTION CALLBACK **********************************************
    
    func onConnected(peripheral: CBPeripheral) {
        callbackToCentral?.onConnected(handler: self)
    }
    
    func onDisconnected(peripheral: CBPeripheral) {
        callbackToCentral?.onDisConnected(handler: self)
    }
    
    func onDataReceived(peripheral: CBPeripheral, data: Data) {
        parser?.push(data: data)
    }
    
    //*************************************************************************************************
}
